### PR TITLE
Implement RSC (lightly tested) / Cleanup hyperparameter registry

### PR DESCRIPTION
**Pull request by SirRob1997**
This PR implements the following features:

- Representation Self-Challenging (RSC) from: https://arxiv.org/abs/2007.02454
- Cleanup algorithm-specific hyperparameters in hyperparameter registry so they don't flood the console by default
- Adapted `MNIST_CNN` forward pass to use `AdaptiveAvgpool2d` instead of mean, this allows for more flexible usage in algorithms and feels more clean

You might also want to delete the MLP parts in `networks.py`  and `hparams_registry.py` as well since I think MLP was only added for Debug datasets that don't get used anymore?

### CODE REVIEW

File: domainbed/algorithms.py [hunk 1]
```diff
@@ -3,10 +3,12 @@
 import copy
 
 import numpy as np
+import numpy.random as npr
```
Comment (lopezpaz) [inline domainbed/algorithms.py:6]: Unused?

File: domainbed/networks.py [hunk 2]
```diff
@@ -5,6 +5,13 @@
 import torch.nn.functional as F
 import torchvision.models
 
+from torch.utils import model_zoo
+from torchvision.models.resnet import BasicBlock, model_urls, Bottleneck
+from torch.autograd import Variable
+import numpy.random as npr
```
Comment (lopezpaz) [inline domainbed/networks.py:11]: Unused?

File: domainbed/networks.py [hunk 3]
```diff
@@ -5,6 +5,13 @@
 import torch.nn.functional as F
 import torchvision.models
 
+from torch.utils import model_zoo
+from torchvision.models.resnet import BasicBlock, model_urls, Bottleneck
+from torch.autograd import Variable
```
Comment (lopezpaz) [inline domainbed/networks.py:10]: Variable has been deprecated from PyTorch.

File: domainbed/networks.py [hunk 4]
```diff
@@ -5,6 +5,13 @@
 import torch.nn.functional as F
 import torchvision.models
 
+from torch.utils import model_zoo
```
Comment (lopezpaz) [inline domainbed/networks.py:8]: Unused?

File: domainbed/networks.py [hunk 5]
```diff
@@ -5,6 +5,13 @@
...
+from torchvision.models.resnet import BasicBlock, model_urls, Bottleneck
+from torch.autograd import Variable
+import numpy.random as npr
+import numpy as np
```
Comment (lopezpaz) [inline domainbed/networks.py:12]: Unused?

File: domainbed/networks.py [hunk 6]
```diff
@@ -5,6 +5,13 @@
...
+import random
```
Comment (lopezpaz) [inline domainbed/networks.py:13]: Many of these imports are unused.

**Comment by lopezpaz** [comment]
For now, I would revert the `hparams_registry.py` cleanup, as this modifies the exact random parameters that are chosen in the overall sweep. @igul222 ?

**Comment by lopezpaz** [comment]
Note: MLP is used in DANN variants.

File: domainbed/algorithms.py (hunk 1 above)
Comment (SirRob1997) [reply domainbed/algorithms.py:6]: Ah yeah I think I have a few imports sill left, my bad

File: domainbed/networks.py (hunk 3 above)
Comment (SirRob1997) [reply domainbed/networks.py:10]: I think these are also still unused leftovers

File: domainbed/networks.py (hunk 4 above)
Comment (SirRob1997) [reply domainbed/networks.py:8]: Yes

**Comment by SirRob1997** [comment]
Then the MLP parameters should only show in the console when DANN is used as an algorithm

**Comment by lopezpaz** [comment]
Merged, although we will have to find a solution to guarantee that the same random parameters are sampled in `hparams_registry` given the same random seed, for all future commits. Right now, adding algorithms or conditional logic in `hparams_registry` changes the random hyperparameters sampled for a given algorithm, dataset, and seed.

cc @igul222

**Comment by lopezpaz** [comment]
@SirRob1997: Preliminar runs show a very poor performance (39%) at OfficeHome. The paper reports 63%. Could you investigate how to get a higher number in this dataset with default hyper-parameters?
